### DISCUSSION TITLE
how to set AlignStart for Box and Entry widgets?

### DISCUSSION THREAD

**Discussion by ITwrx** [can-lehmann/owlkettle#45] [Q&A] [answered]
In the `custom_dialog` example, the Entries' input fields align to the right side of the boxes. Both the containing box and the Property's Entry widget appear to use AlignFill. This looks fine in the example, because the window is small, and all the input fields are the same size. This is not so great when the window needs to be wider due to longer fields, but some fields are still short. 

![custom_dialog](https://user-images.githubusercontent.com/1002762/228033103-4a32eaea-7c5b-48b9-bb5c-244a2262724d.jpg)

My reading of the owlkettle docs says i can set AlignStart for both the box and the property's Entry, but i'm not sure how. I haven't run across an example yet, and i've tried every unholy abomination i can think of, to no avail. :) Any tips appreciated.

**Comment by can-lehmann** [comment, marked as answer]
The entries in `custom_dialog` are right aligned because the labels are set to `expand: true`, while the entries use `expand: false`. This causes the labels to take up any remaining space in the rows.

```nim
Box:
  orient = OrientY
  margin = 12
  spacing = 6
  
  Box {.expand: false.}:
    orient = OrientX
    spacing = 6
    
    Label:
      text = "Some Entry"
      xAlign = 0
    
    Entry() {.expand: false.}
  
  Box {.expand: false.}:
    orient = OrientX
    spacing = 6
    
    Label:
      text = "Entry 2"
      xAlign = 0
    
    Entry() {.expand: false.}
```

![Entries with expand: false](https://user-images.githubusercontent.com/85876381/228064275-5e527c90-4bf7-49f7-a20d-4f79a70b4294.png)

You could set the entries to `expand: true` instead, while using `expand: false` for the labels. This will cause the entries to dynamically take up any remaining space in the rows. However, they are no longer vertically aligned:

![Labels with expand: false](https://user-images.githubusercontent.com/85876381/228064262-661a9b42-b564-4782-89e0-1f9274c4873a.png)

This is probably not what you want though. To me this problem seems like a classic use case for a `Grid`:

```nim
Grid:
  margin = 12
  spacing = 6
  
  Label {.x: 0, y: 0.}:
    text = "Some Entry"
    xAlign = 0
  
  Entry() {.x: 1, y: 0, hExpand: true.}
  
  Label {.x: 0, y: 1.}:
    text = "Entry 2"
    xAlign = 0
  
  Entry() {.x: 1, y: 1, hExpand: true.}
```

The result now looks like this:

![Using a Grid](https://user-images.githubusercontent.com/85876381/228064808-cc67f88c-2f22-426c-b1ad-9c6b2176b2d5.png)

Here it is applied to a case similar to your example, where one entry is wider than the other:

```nim
Grid:
  margin = 12
  spacing = 6
  
  Label {.x: 0, y: 0, hExpand: true.}:
    text = "Some Entry"
    xAlign = 0
  
  Entry(width=50) {.x: 1, y: 0.}
  
  Label {.x: 0, y: 1, hExpand: true.}:
    text = "Entry 2"
    xAlign = 0
  
  Entry() {.x: 1, y: 1.}
```

![Grid with one entry wider than the other](https://user-images.githubusercontent.com/85876381/228065435-17411c96-72b4-4a96-aa98-dfc013cbe49d.png)

If you use this in a larger application, note that you might want to abstract this usage of `Grid` as a "From widget" into a custom `viewable`.

**Reply by ITwrx** [reply]
Thanks a lot for all of that. So, is `AlignStart` really not available for `Box` and `Entry` as the compiler seems to be telling me, because it seems like the docs contradict that. I used gtk inspector and they both showed `AlignFill` and when i changed both to `AlignStart` it was correct for my use case, but i couldn't find a way to change `halign` for either `Box` or `Entry`.

**Reply by can-lehmann** [reply]
The available properties depend on the `adder` of the parent widget you are using (not on the child widget). Here is a really simple example of an Entry that is added to a Box with `hAlign` set to `AlignStart`:

```nim
Box:
  Entry {.hAlign: AlignStart.}:
    discard
```

I would need to see the section of code that causes the error to give a more detailed explanation of what is going on. If you are able to change the value in GTK Inspector, I would expect that you would be able to change it in code too though.

**Reply by ITwrx** [reply]
the code is:

```
import std/strutils

import owlkettle

import ../../models/loan_model

viewable AddLoanDialog:
  loan: Loan

viewable Property:
  name: string
  child: Widget

method view(property: PropertyState): Widget =
  result = gui:
    Box:
      orient = OrientX
      spacing = 6
      
      Label:
        text = property.name
        xAlign = 0
      
      insert(property.child) {.expand: false.}

proc add(property: Property, child: Widget) =
  property.hasChild = true
  property.valChild = child

method view(dialog: AddLoanDialogState): Widget =
  result = gui:
    Dialog:
      title = "New Loan"
      defaultSize = (400, 0)
      
      DialogButton {.addButton.}:
        text = "Create"
        style = {ButtonSuggested}
        res = DialogAccept
      
      DialogButton {.addButton.}:
        text = "Cancel"
        res = DialogCancel
      
      Box:
        orient = OrientY
        spacing = 6
        margin = 12
        
        Property:
          name = "Purchase Item Name:"
          Entry:
            text = ""
            width = 50
            #text = dialog.loan.purchaseItemName
            proc changed(purchaseItemName: string) =
              dialog.loan.purchaseItemName = purchaseItemName

        Property:
          name = "Purchase Item SKU:"
          Entry:
            text = ""
            #text = dialog.loan.purchaseItemSku
            proc changed(purchaseItemSku: string) =
              dialog.loan.purchaseItemSku = purchaseItemSku

        Property:
          name = "Principle:"
          Entry:
            text = ""
            #text = $dialog.loan.principle
            proc changed(principle: string) =
              dialog.loan.principle = parseFloat(principle)

        Property:
          name = "Interest Rate:"
          Entry:
            text = ""
            #text = $dialog.loan.interestRate
            proc changed(interestRate: string) =
              dialog.loan.interestRate = parseFloat(interestRate)      

export AddLoanDialog
```
i get:

```
Error: type mismatch: got <Property, Entry, hAlign: Align>
but expected one of:
proc add(property: Property; child: Widget)

```

if i try to add `{.hAlign: AlignStart.}` to the second Property Entry, for instance. It seems like maybe i should be adding it to `insert(property.child) {.expand: false.}`, but i don't know the syntax for adding two of those.

p.s. i also get an error if i try to use `style = [ButtonSuggested]` like in the docs instead of `style = {ButtonSuggested}` for the Create dialog button, like i have above. idk if that's a bug, or if i'm just doing something wrong. :)

**Reply by can-lehmann** [reply]
In the code you provided, you add an `Entry` to the `Property` using the `add` adder. It does not have any properties (parameters to adders are called properties), especially not `hAlign`. You can add an `hAlign` property if you want to (I use the new adder syntax here, instead of defining the adder as a proc):

```nim
viewable Property:
  name: string
  child: (Align, Widget)
  
  adder add {.hAlign: AlignFill.}:
    widget.hasChild = true
    widget.valChild = (hAlign, child)
```

You can then change the `view` method to use the passed property. I don't really see how exactly you want to change the `hAlign` to fix your issue, so here is just one example:

```nim
method view(property: PropertyState): Widget =
  result = gui:
    Box:
      orient = OrientX
      spacing = 6
      
      Label:
        text = property.name
        xAlign = 0
      
      insert(property.child[1]) {.expand: false, hAlign: property.child[0].}
```

For more on adders check out the documenation here: https://can-lehmann.github.io/owlkettle/book/internals/adders.html.

Re: `style = ...`: The documentation is built from the development version of owlkettle, so it already includes any changes made that are not yet published as part of a numbered version. The syntax for styling was changed in the development version to allow for custom CSS classes: https://github.com/can-lehmann/owlkettle/pull/40

**Reply by can-lehmann** [reply]
I just read my reply again and noticed that it is a bit confusing because the widget you use is also called `Property`. So to clarify: When I write `Property`, I mean the custom widget you used in your code. When I write "property"/"properties" I refer to parameters to an adder as explained here: https://can-lehmann.github.io/owlkettle/book/internals/adders/adders_with_properties.html

**Reply by ITwrx** [reply]
no problem. i caught that. thanks though.

**Reply by ITwrx** [reply]
> I use the new adder syntax here, instead of defining the adder as a proc

Saboteur!!! j/k
Is that supported in 2.2.0, because i'm just getting `undeclared identifier: adder` if i blindly copy/paste your adder to try this out. :) I haven't figured out how to add a property to a proc adder... :(

**Reply by can-lehmann** [reply]
I am pretty sure that it is supported in `2.2.0`. Here is the entire code I used to test the adder (based on the code you provided). Note again that as I don't currently see where you want to actually use the `hAlign: AlignStart`, this version does not fix the original issue:

```nim
import std/strutils

import owlkettle

#import ../../models/loan_model

viewable AddLoanDialog:
  discard #loan: Loan

viewable Property:
  name: string
  child: (Align, Widget)
  
  adder add {.hAlign: AlignFill.}:
    widget.hasChild = true
    widget.valChild = (hAlign, child)

method view(property: PropertyState): Widget =
  result = gui:
    Box:
      orient = OrientX
      spacing = 6
      
      Label:
        text = property.name
        xAlign = 0
      
      insert(property.child[1]) {.expand: false, hAlign: property.child[0].}

method view(dialog: AddLoanDialogState): Widget =
  result = gui:
    Dialog:
      title = "New Loan"
      defaultSize = (400, 0)
      
      DialogButton {.addButton.}:
        text = "Create"
        style = [ButtonSuggested]
        res = DialogAccept
      
      DialogButton {.addButton.}:
        text = "Cancel"
        res = DialogCancel
      
      Box:
        orient = OrientY
        spacing = 6
        margin = 12
        
        Property:
          name = "Purchase Item Name:"
          Entry:
            text = ""
            width = 50
            #text = dialog.loan.purchaseItemName
            #proc changed(purchaseItemName: string) =
            #  dialog.loan.purchaseItemName = purchaseItemName

        Property:
          name = "Purchase Item SKU:"
          Entry {.hAlign: AlignStart.}:
            text = ""
            #text = dialog.loan.purchaseItemSku
            #proc changed(purchaseItemSku: string) =
            #  dialog.loan.purchaseItemSku = purchaseItemSku

        Property:
          name = "Principle:"
          Entry:
            text = ""
            #text = $dialog.loan.principle
            #proc changed(principle: string) =
            #  dialog.loan.principle = parseFloat(principle)

        Property:
          name = "Interest Rate:"
          Entry:
            text = ""
            #text = $dialog.loan.interestRate
            #proc changed(interestRate: string) =
            #  dialog.loan.interestRate = parseFloat(interestRate)      

brew(gui(AddLoanDialog()))
```

**Reply by ITwrx** [reply]
thanks a lot. i'll check it all out later/tomorrow and report back.

**Reply by ITwrx** [reply]
Re-reading your answers, i just changed the `Label:` to `{.expand: false.}` which gives me this:

![label_expand_false](https://user-images.githubusercontent.com/1002762/228110789-700a1daf-a75e-40ba-bcf4-af5a7d83b200.png)

which is good enough for my purposes right now. I'm really just trying to build a somewhat realistic CRUD POC to learn and work from going forward. 

> Note again that as I don't currently see where you want to actually use the hAlign: AlignStart, this version does not fix the original issue:

I wanted to add it to every Property's Entry and to the containing box, but since gtk won't let me add a property to the first widget in a gui tree, i was going to add a second box. I thought i had tried `expand: false` on the `label` before, so `AlignStart` was my next attempt at just getting the entries over by the labels. I'm not too worried about vertical alignment right now.

Thanks for showing me how to add properties to adders. I didn't get it working immediately, and gave up on that for now, but I'm sure i'll need that soon enough and will try again then, but with this info, and a maybe re-read of the owlkettle book section on adders.
